### PR TITLE
Add simple downhill water movement timestep

### LINKED ISSUE #9: Add simple downhill water movement timestep
Add the first placeholder water movement implementation inside the simulation timestep.

Scope:
- Compare water-surface elevation (`terrain + depth`) against four neighboring cells.
- Move water from a wet cell toward lower neighboring cells using hard-coded simple routing values.
- Treat out-of-domain and invalid/nodata neighbors as open drains so water can leave the domain.
- Keep this intentionally simple for now, without inertia, Manning friction, or finite-volume face momentum.
- Add tests for lower-neighbor routing, open-boundary drainage, and invalid-cell drainage.

### PR DESCRIPTION
**Pull request by richpsharp**
## Summary

Closes #9.

Adds the first intentionally simple water movement implementation to the simulation timestep. The new `water_timestep` compares cell water-surface elevation (`terrain + depth`) against four cardinal neighbors and moves a hard-coded capped fraction of water toward lower receivers.

Out-of-domain and invalid/nodata neighbors are treated as open low drains, so water can leave the raster when it reaches a boundary or undefined pixel. This does not yet model inertia, Manning friction, face momentum, or CFL-limited finite-volume fluxes.

## Validation

- `python -m compileall src tests run_model.py`
- `python -m unittest discover -s tests`

### CODE REVIEW

File: src/nefas/engine.py [hunk 1]
```diff
@@ -6,21 +6,19 @@
...
-matplotlib.use("Agg")
-from matplotlib import pyplot as plt
 
 from .config import RainfallConfig, SimulationConfig
-from .preprocessing import IntermediateOutputs
 from .simulation import RasterGrid, SimulationState
 
+if TYPE_CHECKING:
+    from .preprocessing import IntermediateOutputs
+
```
Comment (richpsharp) [inline src/nefas/engine.py]: Just do the imports directly always, not conditional

File: src/nefas/engine.py [hunk 2]
```diff
@@ -145,6 +143,10 @@ def timestep_duration_seconds(
 
 def storm_footprint_mask(storm_footprint: Path, template_raster: Path) -> np.ndarray:
     """Rasterize the storm footprint onto the clipped DEM grid."""
+    import geopandas as gpd
+    import rasterio
+    from rasterio.features import rasterize
+
```
Comment (richpsharp) [inline src/nefas/engine.py]: import at the top of the file

File: src/nefas/engine.py [hunk 3]
```diff
@@ -252,8 +254,133 @@ def add_rainfall_depth(
...
+    hydraulic.qx[:, 1:-1] = local_inertial_flux_update(
+        old_flux=qx_old,
+        face_depth=h_face_x,
+        slope=slope_x,
+        manning_n=n_face_x,
+        valid_faces=valid_x,
+        dt_seconds=dt_seconds,
+    )
+    hydraulic.qx[:, 0] = 0
+    hydraulic.qx[:, -1] = 0
```
Comment (richpsharp) [inline src/nefas/engine.py]: i would like an open boundary condition here, so boundary is treated as another cell but the flux just disappears

File: src/nefas/engine.py [hunk 4]
```diff
@@ -252,8 +254,133 @@ def add_rainfall_depth(
...
+    hydraulic.qy[1:-1, :] = local_inertial_flux_update(
+        old_flux=qy_old,
+        face_depth=h_face_y,
+        slope=slope_y,
+        manning_n=n_face_y,
+        valid_faces=valid_y,
+        dt_seconds=dt_seconds,
+    )
+    hydraulic.qy[0, :] = 0
+    hydraulic.qy[-1, :] = 0
```
Comment (richpsharp) [inline src/nefas/engine.py]: Same here for boundary condition

File: src/nefas/engine.py [hunk 5]
```diff
@@ -252,8 +254,133 @@ def add_rainfall_depth(
...
+    )
+
+
+def face_depth_y(elevation: np.ndarray, water_surface: np.ndarray) -> np.ndarray:
+    """Return representative depths at north-south interior faces."""
+    return np.maximum(
+        0,
+        np.maximum(water_surface[:-1, :], water_surface[1:, :])
+        - np.maximum(elevation[:-1, :], elevation[1:, :]),
+    )
```
Comment (richpsharp) [inline src/nefas/engine.py]: These are both thin functions and used only once, just include the code inline rather than calling it and leave a comment saying what it does

File: src/nefas/engine.py [hunk 6]
```diff
@@ -295,6 +422,11 @@ def render_snapshot(
     elapsed_minutes: float,
 ) -> None:
     """Render DEM, storm footprint, and water depth into a PNG image."""
+    import matplotlib
+
+    matplotlib.use("Agg")
+    from matplotlib import pyplot as plt
```
Comment (richpsharp) [inline src/nefas/engine.py]: import at the top

File: src/nefas/simulation.py [hunk 7]
```diff
@@ -56,6 +55,8 @@ def nx(self) -> int:
     @classmethod
     def from_dem(cls, path: Path) -> RasterGrid:
         """Build a terrain grid from the first band of a DEM raster."""
+        import rasterio
```
Comment (richpsharp) [inline src/nefas/simulation.py]: import at top

**Review by richpsharp — COMMENTED** [review]
address comments

File: src/nefas/engine.py (hunk 1 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [e970269](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/e970269c9910346c63277a6e45d46d232725ccba): imports are back at module scope instead of conditional/type-checking-only.

File: src/nefas/engine.py (hunk 2 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [e970269](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/e970269c9910346c63277a6e45d46d232725ccba): moved the geospatial imports back to module scope.

File: src/nefas/engine.py (hunk 3 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [05628cc](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/05628cc63b2500ab6c9c0775986d20f48051d99c): x-boundary faces now compute one-way outward flux against a dry outside cell, so boundary outflow leaves the domain.

File: src/nefas/engine.py (hunk 4 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [05628cc](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/05628cc63b2500ab6c9c0775986d20f48051d99c): y-boundary faces now use the same one-way open-boundary outflow treatment.

File: src/nefas/engine.py (hunk 5 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [042597e](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/042597e4f0785bfea9662a262acb84f4eaeb3b05): inlined both face-depth calculations and left a short comment describing the sill-depth logic.

File: src/nefas/engine.py (hunk 6 above)
Comment (richpsharp) [reply src/nefas/engine.py]: Fixed in [e970269](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/e970269c9910346c63277a6e45d46d232725ccba), with backend ordering preserved in [4698c82](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/4698c8298e769ecfe9fdee1af1957a0869fb566a): the matplotlib imports are module-scope, with `pyplot` loaded after selecting the Agg backend.

File: src/nefas/simulation.py (hunk 7 above)
Comment (richpsharp) [reply src/nefas/simulation.py]: Fixed in [e970269](https://github.com/springinnovate/natural-ecosystem-flood-attenuation-simulator/commit/e970269c9910346c63277a6e45d46d232725ccba): `rasterio` is imported at module scope again.

**Comment by richpsharp** [comment]
Proposal for fixed snapshot depth scale:

The snapshot renderer should use a user-configurable maximum water-depth scale so colors remain comparable across frames. Right now each PNG can effectively rescale to the current depth range, which makes it hard to tell whether flooding is intensifying or receding.

Suggested config shape:

```yaml
output:
  directory: outputs/example
  snapshots:
    directory: snapshots
    interval_minutes: 15
    max_depth_meters: 1.0
```

Implementation outline:

1. Add `max_depth_meters: float | None = None` to `SnapshotConfig`.
2. Parse optional `output.snapshots.max_depth_meters` as a positive float.
3. Thread the value through `write_snapshot(...)` into `render_snapshot(...)`.
4. In `render_snapshot`, pass fixed `vmin=0` and `vmax=max_depth_meters` to the water-depth `imshow` when configured.
5. Keep current auto-scaling behavior when `max_depth_meters` is omitted.
6. Add config parsing tests and a renderer/unit test that verifies configured `vmax` is used.

This keeps the default behavior backward-compatible while letting screening runs produce frame sequences with stable, comparable color meaning.
